### ISSUE TITLE
Measure dense OBB assignment memory and fallback equivalence

### ISSUE THREAD

**Issue by shapovalovei** [shapovalovei/YOLOX-OBB#27]
## Phase

Performance/correctness validation only. This Issue does not authorize implementation.

## Handoff source

Issue #23 research identified dense assignment memory as a potentially useful improvement, but no local defect or supported-workload OOM was proven. Orchestrator review intentionally separated this from non-OOM RuntimeError classification tracked elsewhere.

## Goal

Measure the maintained OBB assignment path under deterministic synthetic dense-target workloads and determine whether peak memory, latency, or CPU-fallback numerical differences justify a later optimization phase.

## Current local behavior

The assignment path materializes multiple tensors proportional to `[num_gt, num_anchor]`, including rotated candidate masks, pairwise KLD, repeated class targets/predictions, costs, and matching state. On RuntimeError the current implementation can retry assignment on CPU.

## Required validation

No model training or dataset changes.

Use synthetic deterministic inputs representative of maintained 416-style anchor counts and sweep target density over bounded values. Where an already-available CUDA environment permits, record:

- candidate anchor counts;
- peak allocated/reserved memory attributable to assignment where measurable;
- assignment runtime after warm-up;
- whether/when OOM occurs;
- GPU result vs CPU-fallback assignment equivalence for matched indices/classes/scores/masks;
- effect of dtype/autocast where already supported;
- scaling with `num_gt` and candidate-anchor count.

CPU-only structural/memory evidence may be collected when CUDA is unavailable, but do not extrapolate GPU peak memory or AMP equivalence from CPU results.

## Decision

Choose one:

A — current assignment resource behavior is acceptable for bounded representative workloads; no optimization phase justified.

B — deterministic supported-scale workloads show material avoidable peak memory/latency or OOM risk; optimization validation/implementation is justified.

C — CPU fallback produces materially different assignment outputs under a maintained dtype/autocast path; correctness issue identified.

D — insufficient environment/evidence.

## Ownership gate

Do not optimize inside this Issue. Any future implementation must be independently derived from local YOLOX-OBB behavior and preserve assignment semantics.

## Non-goals

- no tracked changes;
- no training;
- no dataset work;
- no assignment algorithm redesign;
- no CI work;
- no external-code copying.

## Completion

Post a factual handoff with base SHA, environment, synthetic matrix, measurements/equivalence results, FACT/INFERENCE/UNVERIFIED separation, verdict, and recommended next phase. Do not close the Issue; owner/orchestrator reviews and closes it.

**Comment by shapovalovei** [comment]
## Validation handoff

### Base

* Repository: `shapovalovei/YOLOX-OBB`
* Fetched `origin/main`: `0695e2834ff5716dfd51fcd6ac7b2300b7bc27ce`
* The requested SHA is an ancestor of `origin/main` (and is the exact fetched `origin/main`).
* The existing checkout was clean on `fix/30-obb-mixup-boundary-geometry` at `9863c167d87ec2f99f6561334c199bd7c918d96c`; its tree is identical to `origin/main`. No branch, source change, test change, commit, or PR was created.
* Environment: macOS 26.6.2 x86_64, 16 GiB RAM, Python 3.11.11 from `/Users/eugene/creditCardReader/venv`, PyTorch `2.2.2`, `torch.version.cuda=None`, `torch.backends.cuda.is_built()=False`.
* `torch.cuda.is_available()=False`, `torch.cuda.device_count()=0`. MPS was not substituted for CUDA.

### Maintained assignment path

The inspected path is `YOLOXHeadOBB_KLD.get_assignments`, `get_in_boxes_info`, and `dynamic_k_matching` from the exact fetched tree.

* Default `mode="gpu"` was measured on this CPU-only build; it operates on the supplied CPU tensors and does not exercise a CUDA device.
* `mode="cpu"` explicitly converts assignment inputs to CPU/FP32, then returns four tensors with `.cuda()`. Consequently, the maintained CPU fallback cannot complete end-to-end on this CPU-only build; a real fallback-equivalence verdict was not attempted or inferred.
* Candidate gating creates repeated `[G,A]` center/delta/boundary tensors, `[G,A,4]` rotated-box and center-delta tensors, boolean masks, and final `[G,K]` candidate state.
* Assignment creates pairwise KLD and approximate-IoU `[G,K]`, repeated target/prediction class state `[G,K,C]`, BCE class loss `[G,K,C]`, cost `[G,K]`, and `matching_matrix [G,K]`.

### Synthetic workload

* Anchors were constructed exactly in maintained P3/P4/P5 order with strides 8/16/32. At 416: `52^2 + 26^2 + 13^2 = 3549`; at 1024: `128^2 + 64^2 + 32^2 = 21504`.
* Every cell used finite positive deterministic OBBs, deterministic angles, deterministic non-degenerate predictions, and a fixed seed (`2701`). No checkpoint or dataset was loaded.
* `S` = image-spanning small/moderate rotated boxes; `D` = large overlapping rotated boxes covering the lattice. K was recorded from the live `get_in_boxes_info` result.
* The 416 primary matrix covers `G=1,5,10,25,50,100,150` for `C=1` and `C=16`, in both regimes. The bounded 1024 structural reference covers `G=1,10,50`, `C=16`, in both regimes. The 1024 `G=400,C=16` upper-capacity cell was not run after analytic estimation (about 2.5 GiB of phase-sum payload before allocator overhead).

### Tensor byte accounting

All primary measurements are FP32 (`B=4` bytes). Exact shape terms are:

* each `[G,A]` float tensor: `4GA` bytes;
* each `[G,A,4]` float tensor: `16GA` bytes;
* each `[G,K]` float tensor: `4GK` bytes;
* each `[G,K,C]` float tensor: `4GKC` bytes;
* boolean masks use one byte per element in this accounting.

The candidate-gate census has 22 `[G,A]`-equivalent float materializations across centers, deltas, four bounds, and two `[G,A,4]` stacks, plus masks. The assignment phase has the `[G,K]` KLD/cost/matching terms and repeated `[G,K,C]` class terms. The `approx major-live` column below is an analytic phase-sum envelope using `22*[G,A]` floats, up to three `[G,K,C]` floats, up to seven `[G,K]` floats, and boolean masks. It is not a measured peak and must not be read as a claim that every term is simultaneously live.

### CPU memory and latency

Each cell ran in a fresh subprocess for RSS isolation. `resource.getrusage().ru_maxrss` was detected as bytes on Darwin. Timing used 3 warmups and 5 timed calls with `time.perf_counter`; the table shows the median. `Delta/peak RSS` is assignment-call-attributable post-input delta / process peak; process startup/import and input storage remain in the absolute peak.

| size | C | G | regime | A | K | largest individual | approx major-live | median ms | Delta/peak RSS MiB |
|---:|---:|---:|:--:|---:|---:|:--|---:|---:|---:|
|416|1|1|S|3549|76|[G,A,4] 0.05|0.31|2.4|2.7/165.3|
|416|1|1|D|3549|3549|[G,A,4] 0.05|0.45|3.7|3.5/165.9|
|416|1|5|S|3549|375|[G,A,4] 0.27|1.60|7.9|3.7/166.0|
|416|1|5|D|3549|3549|[G,A,4] 0.27|2.22|12.5|4.3/166.5|
|416|1|10|S|3549|680|[G,A,4] 0.54|3.32|16.4|4.8/167.1|
|416|1|10|D|3549|3549|[G,A,4] 0.54|4.44|25.1|7.0/169.3|
|416|1|25|S|3549|1440|[G,A,4] 1.35|9.03|50.8|9.5/171.6|
|416|1|25|D|3549|3549|[G,A,4] 1.35|11.09|84.8|9.7/172.1|
|416|1|50|S|3549|2044|[G,A,4] 2.71|19.23|154.7|19.2/181.7|
|416|1|50|D|3549|3549|[G,A,4] 2.71|22.18|206.0|19.5/182.1|
|416|1|100|S|3549|2853|[G,A,4] 5.42|41.62|456.9|35.9/198.3|
|416|1|100|D|3549|3549|[G,A,4] 5.42|44.34|554.2|36.3/198.9|
|416|1|150|S|3549|2789|[G,A,4] 8.12|62.06|1076.8|53.7/216.4|
|416|1|150|D|3549|3549|[G,A,4] 8.12|66.51|1375.5|54.5/217.2|
|416|16|1|S|3549|76|[G,A,4] 0.05|0.33|9.4|2.4/165.3|
|416|16|1|D|3549|3549|[G,K,C] 0.22|1.06|24.6|3.8/167.0|
|416|16|5|S|3549|375|[G,A,4] 0.27|1.93|32.3|3.8/166.5|
|416|16|5|D|3549|3549|[G,K,C] 1.08|5.27|56.5|8.4/171.2|
|416|16|10|S|3549|680|[G,A,4] 0.54|4.49|62.3|5.6/168.4|
|416|16|10|D|3549|3549|[G,K,C] 2.17|10.53|99.7|13.8/176.9|
|416|16|25|S|3549|1440|[G,K,C] 2.20|15.21|131.1|12.1/174.8|
|416|16|25|D|3549|3549|[G,K,C] 5.42|26.32|192.9|26.4/188.8|
|416|16|50|S|3549|2044|[G,K,C] 6.24|36.78|291.9|28.5/191.0|
|416|16|50|D|3549|3549|[G,K,C] 10.83|52.64|395.0|49.2/211.8|
|416|16|100|S|3549|2853|[G,K,C] 17.41|90.60|763.9|74.2/236.7|
|416|16|100|D|3549|3549|[G,K,C] 21.66|105.27|822.0|92.3/254.9|
|416|16|150|S|3549|2789|[G,K,C] 25.53|133.87|1205.2|106.5/269.1|
|416|16|150|D|3549|3549|[G,K,C] 32.49|157.90|1329.5|134.8/297.5|
|1024|16|1|S|21504|75|[G,A,4] 0.33|1.90|10.3|1.7/169.6|
|1024|16|1|D|21504|21504|[G,K,C] 1.31|6.42|36.9|7.9/176.1|
|1024|16|10|S|21504|793|[G,A,4] 3.28|20.17|61.2|20.4/188.2|
|1024|16|10|D|21504|21504|[G,K,C] 13.12|63.82|294.6|49.5/217.3|
|1024|16|50|S|21504|3602|[G,A,4] 16.41|130.28|479.3|104.8/272.5|
|1024|16|50|D|21504|21504|[G,K,C] 65.62|318.94|1427.9|232.7/400.6|

### PyTorch profiler

Profiler was available without setup and was run for 416/D/G=25/C=16. It reported cumulative self allocations of approximately:

* `aten::empty_strided`: 31.34 MB, 3,333 calls;
* `aten::where`: 25.55 MB, 1,675 calls;
* `aten::mul`: 16.25 MB self, 16.60 MB including children;
* `aten::cat`: 9.65 MB, 54 calls;
* `aten::repeat`: 15.62 MB cumulative CPU memory attributed, 34 calls, with 0.14 MB direct self allocation;
* `aten::binary_cross_entropy`: 5.68 MB cumulative output memory;
* `aten::min`: 2.13 MB and `aten::topk`: 3.9 KB.

These are cumulative profiler allocation/accounting values, not peak RSS. The KLD guarded elementwise operations and repeated concatenation account for substantial allocation activity; the byte census identifies the large shape-level class/geometry materializations.

### CUDA memory

**UNVERIFIED.** CUDA is not built or available locally. No CUDA allocator setup, device setup, or artificial allocation was attempted. CUDA allocated/reserved peaks and CUDA OOM threshold are unavailable.

### CUDA latency

**UNVERIFIED.** No CUDA timing was attempted because the availability gate failed.

### OOM behavior

**UNVERIFIED for CUDA.** No bounded cell could naturally raise `torch.cuda.OutOfMemoryError` in this environment. No workload was increased to force an OOM. The accepted #40/#41 classification boundary was not reopened or revalidated here.

### GPU/CPU fallback equivalence

**UNVERIFIED.** Genuine CUDA comparison was impossible. The CPU-only maintained `mode="cpu"` return path reaches `.cuda()` and cannot complete on this build. No monkeypatch was used and no CPU-only result was presented as fallback equivalence.

### Dtype/autocast

* FP32 is the required measured matrix dtype.
* Direct BF16 and FP16 input probes at 416/D/G=25/C=16 completed on CPU. Pairwise KLD was `torch.float32`; returned assignment tensors were `float32`, `bool`, `float32`, and `int64` in the expected positions. Median times were 94.1 ms (BF16 inputs) and 91.8 ms (FP16 inputs); these are diagnostic CPU input-dtype results, not CUDA AMP qualification.
* Existing CPU autocast contexts for BF16 and FP16 completed the same deterministic 416/D/G=25/C=16 call. Pairwise KLD and assignment outputs remained FP32/boolean/int64 as above. No CUDA autocast evidence exists.
* This matches current source behavior: class predictions and one-hot targets are explicitly converted to FP32, CPU mode converts several inputs to FP32, and KLD internally promotes half/BF16 arithmetic to FP32.

### Scaling analysis

* **FACT:** the candidate gate has `O(G*A)` work/storage. Dense 416 keeps `K=A=3549`; sparse 416 measured `K=76,375,680,1440,2044,2853,2789` for G=1…150, so candidate filtering changes the observed cost and saturates with overlapping image-spanning placements.
* **FACT:** pairwise KLD, class cost, and dynamic matching scale with `G*K`; class-target/prediction/BCE materializations scale with `G*K*C`.
* **FACT:** for dense 416/C=16, median CPU time grew from 192.9 ms at G=25 to 395.0 ms at G=50, 822.0 ms at G=100, and 1329.5 ms at G=150. Delta RSS grew from 26.4 to 49.2, 92.3, and 134.8 MiB.
* **FACT:** at 1024/D/G=50/C=16, increasing A from 3549 to 21504 and K from 3549 to 21504 increased median time from 395.0 to 1427.9 ms and delta RSS from 49.2 to 232.7 MiB.
* **INFERENCE:** both target density/candidate count and class count are meaningful resource dimensions; a future lower-materialization design would need semantic assignment-equivalence proof. This validation did not benchmark such a design.

### Resource hotspots

1. **FACT:** `[G,K,C]` repeated class targets/predictions/BCE state is the largest individual tensor for dense 416/C=16 from G=1 onward except the smallest sparse cells; at dense 416/G=150 it is 32.49 MiB per tensor.
2. **FACT:** `[G,A,4]` rotated candidate/center-delta tensors are the largest individual FP32 tensor for C=1 and for sparse low-K cells; at 416/G=150 it is 8.12 MiB.
3. **FACT:** `[G,K]` KLD/approximate-IoU/cost/matching terms multiply with G and K and are visible in both the byte census and profiler allocation activity.
4. **INFERENCE:** `repeat` is a plausible future optimization target because it materializes the stated shapes, but its appearance alone does not prove that an alternative broadcast/chunked formulation is semantically safe or lower peak.

### Existing tests

Required focused commands passed:

* `python -m unittest discover -s tests -p 'test_assignment_fallback.py' -v`: 8/8
* `python -m unittest discover -s tests -p 'test_obb_assignment_geometry.py' -v`: 5/5
* `python -m unittest discover -s tests -p 'test_kld_prediction_target_contract.py' -v`: 5/5
* `python -m unittest discover -s tests -p 'test_kld_numerical_stability.py' -v`: 18/18
* `python -m unittest discover -s tests -p 'test_obb_target_validity.py' -v`: 11/11

Broader command `python -m unittest discover -s tests -p 'test_*.py' -v` reached 101 tests: 93 passed, 6 skipped, and 2 import errors. The errors were only `test_numpy2_rotated_postprocess` and `test_polyiou_packaging`, both due to missing optional native `DOTA_devkit_YOLO._polyiou`; no assignment or required neighboring OBB test failed. The six skips were the existing ONNX/model-dependent skips caused by the same missing native import.

### FACT / INFERENCE / UNVERIFIED

* **FACT:** exact 416 anchor count is 3549 and exact 1024 structural count is 21504.
* **FACT:** the deterministic primary CPU matrix completed through G=150 for C=1 and C=16 in sparse and dense regimes without natural CPU OOM.
* **FACT:** measured CPU RSS delta/peak for the primary matrix reached 134.8/297.5 MiB at 416/D/G=150/C=16.
* **FACT:** assignment latency and CPU resource use increase with G and K; C=16 materially increases the measured high-G RSS envelope over C=1.
* **FACT:** accepted #40/#41 fallback classification remains in the inspected source; this phase did not alter or broaden that policy.
* **INFERENCE:** dense repeated class state and KLD/cost allocation activity are plausible future optimization targets, but no alternative was implemented or measured.
* **UNVERIFIED:** CUDA peak allocation/reservation, natural CUDA OOM threshold, synchronized CUDA latency, GPU/CPU fallback equivalence, near-tie GPU/CPU assignment behavior, CUDA AMP behavior, and any model-quality/training impact.

### Verdict

**D — INSUFFICIENT EVIDENCE.**

The CPU evidence establishes bounded assignment scaling and identifies concrete materializations, but this phase cannot justify an A/B/C decision for the CUDA OOM boundary or GPU-to-CPU fallback equivalence because CUDA is unavailable. The CPU results alone are not translated into GPU VRAM or CUDA OOM claims.

### Recommended next phase

Do not create an implementation Issue from this handoff. If the owner wants to resolve the remaining decision, repeat a small subset on an already-available CUDA environment: 416/D and 416/S at G=25,100,150 for C=1/16, plus 1024/D/G=50/C=16; collect synchronized CUDA allocator/timing evidence, natural bounded OOM behavior only, and genuine `mode="gpu"` versus `mode="cpu"` output comparisons for stable non-tie and near-tie fixtures. Missing CUDA evidence is what could change this verdict to A, B, or C.

No source, test, algorithm, fallback policy, training, dataset, CI, branch, commit, PR, or Issue closure was performed.

**Comment by shapovalovei** [comment]
## Orchestrator acceptance

Validation accepted with **Verdict D — insufficient evidence**.

The phase established useful bounded CPU evidence without over-claiming CUDA behavior:

- exact 416 anchor contract (`3549`) and bounded 1024 structural reference;
- deterministic sparse/dense GT matrices across `G` and `C`;
- measured CPU RSS/latency scaling and tensor-byte census;
- evidence-ranked `[G,K,C]`, `[G,A,4]`, and `[G,K]` resource hotspots;
- no natural CPU OOM in the bounded matrix;
- no CUDA setup or forced OOM;
- no false GPU-memory extrapolation from CPU RSS;
- no monkeypatched CPU-only result presented as maintained GPU↔CPU fallback equivalence.

CUDA peak memory, CUDA OOM threshold, CUDA latency/AMP, and true GPU↔CPU fallback equivalence remain **UNVERIFIED** because CUDA is unavailable in the validation environment. Therefore the evidence does not justify an assignment optimization implementation phase now.

No implementation Issue should be created from #27 alone. If a suitable CUDA environment becomes naturally available later, selected 416/1024 cells can be revalidated as a separate follow-up.

This completes the bounded validation scope of #27. The framework audit can now proceed to final downstream handoff documentation in #28.
